gap/attr.gi

```diff
@@ -17,85 +17,152 @@ InstallMethod(OutNeighbours, "for a digraph by out-neighbours",
+      counter := counter + 1;
+      pre[v]  := counter;
+      low[v]  := counter;
     fi;
-    i := PositionProperty(nbs[v], w -> w <> v, stack[depth][2]);
+    i := PositionProperty(nbs[v], w -> w <> v, i);
     while i <> fail do
       w := nbs[v][i];
-      if not visited[w] then
-        parent[w]       := v;
-        stack[depth][2] := i;
-        depth           := depth + 1;
-        if not IsBound(stack[depth]) then
-          stack[depth] := [];
+      if pre[w] <> -1 then
+        # v -> w is a back edge
+        if w <> u and pre[w] < low[v] then
+          low[v] := pre[w];
+        fi;
+        orientation[v][w] := not orientation[w][v];
+        i := PositionProperty(nbs[v], w -> w <> v, i);
+      else
+        # diving - part 2
+        if v = 1 then
+          nr_children := nr_children + 1;
         fi;
-        stack[depth][1] := w;
-        stack[depth][2] := 0;
+        orientation[v][w] := true;
+        Push(stack, u);
+        Push(stack, v);
+        Push(stack, i);
+        u := v;
+        v := w;
+        i := 0;
         break;
-      elif parent[v] <> w and num[w] < low[v] then
-        low[v] := num[w];
       fi;
-      i := PositionProperty(nbs[v], w -> w <> v, i);
     od;
-  od;
+  until Size(stack) = 0;
 
   if counter = DigraphNrVertices(D) then
-    i := Position(parent, 1, 1);
-    if i <> fail and Position(parent, 1, i) <> fail then
-      Add(points, 1);
+    connected := true;
+    if nr_children > 1 then
+      Add(articulation_points, 1);
     fi;
-    if IsImmutableDigraph(D) then
-      SetIsConnectedDigraph(D, true);
+    if not IsEmpty(bridges) then
+      orientation := fail;
+    else
+      orientation := DigraphByAdjacencyMatrix(DigraphMutabilityFilter(D),
+                                              orientation);
     fi;
-    return points;
   else
-    if IsImmutableDigraph(D) then
-      SetIsConnectedDigraph(D, false);
+    connected           := false;
+    articulation_points := [];
+    bridges             := [];
+    orientation         := fail;
+  fi;
+  if IsImmutableDigraph(D) then
+    SetIsConnectedDigraph(D, connected);
+    SetArticulationPoints(D, articulation_points);
+    SetBridges(D, bridges);
+    if IsSymmetricDigraph(D) then
+      SetStrongOrientationAttr(D, orientation);
     fi;
-    return [];
   fi;
+  return [connected, articulation_points, bridges, orientation];
+end);
+
+InstallMethod(ArticulationPoints, "for a digraph by out-neighbours",
+[IsDigraphByOutNeighboursRep],
+function(D)
+  return DIGRAPHS_ArticulationPointsBridgesStrongOrientation(D)[2];
+end);
+
+InstallMethod(Bridges, "for a digraph by out-neighbours",
+[IsDigraphByOutNeighboursRep],
+function(D)
+  return DIGRAPHS_ArticulationPointsBridgesStrongOrientation(D)[3];
+end);
+
+InstallMethodThatReturnsDigraph(StrongOrientation,
+"for a digraph by out-neighbours",
+[IsDigraphByOutNeighboursRep],
+function(D)
+  if not IsSymmetricDigraph(D) then
+    ErrorNoReturn("not yet implemented");
```

Comment:
Is there any prospect of a notion of `StrongOrientation` for non-symmetric digraphs? This error message would suggest that there is a notion, we just haven't going round to programming it yet. Which might set the wrong expectation.

Reply:
Yes there is, I just didn't have time to write the code for it. The notion certainly makes sense, for every pair of edges u -> v and v -> u  in a digraph, then we'd be asking to choose one of them so that the resulting digraph had no such pairs and was strongly connected.